testsuite/tests/typing-abstract-kinds/basics.ml

```diff
@@ -0,0 +1,766 @@
+
+module C = struct
+  kind_ kc = B.kb
+end
+
+module D = struct
+  kind_ kd = C.kc mod global
+end
+
+type t : D.kd
+
+type s1 : float64 mod portable global = t
+type s2 : any mod portable global = t
+[%%expect{|
+module A : sig kind_ ka = float64 end
+module B : sig kind_ kb = float64 mod portable end
+module C : sig kind_ kc = float64 mod portable end
+module D : sig kind_ kd = float64 mod global portable end
+type t : float64 mod global portable
+type s1 = t
+type s2 = t
+|}]
+
+type s3 : any mod contended = t
+[%%expect{|
+Line 1, characters 0-31:
+1 | type s3 : any mod contended = t
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The kind of type "t" is float64 mod global portable
+         because of the definition of t at line 17, characters 0-13.
+       But the kind of type "t" must be a subkind of any mod contended
+         because of the definition of s3 at line 1, characters 0-31.
+|}]
+
+(************************)
+(* Test: no with bounds *)
+
+type t
+
+kind_ k = immutable_data with t
+[%%expect{|
+type t
+Line 3, characters 30-31:
+3 | kind_ k = immutable_data with t
+                                  ^
+Error: 'with' syntax is not allowed in kind declarations.
+|}]
+
+(**********************)
+(* Test: no recursion *)
+
+kind_ k_rec = k_rec
+[%%expect{|
+Line 1, characters 14-19:
+1 | kind_ k_rec = k_rec
+                  ^^^^^
+Error: Unbound jkind "k_rec"
+|}]
+
+(***************************************)
+(* Test: you can shadow built-in kinds *)
+
+module Spicy = struct
+  kind_ value = float64 & float64
+
+  type t : value = #(float# * float#)
+end
+[%%expect{|
+module Spicy :
+  sig kind_ value = float64 & float64 type t = #(float# * float#) end
+|}]
+
+type t : Spicy.value = string
+[%%expect{|
+Line 1, characters 0-29:
+1 | type t : Spicy.value = string
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type "string" is value
+         because it is the primitive type string.
+       But the layout of type "string" must be a sublayout of float64 & float64
+         because of the definition of t at line 1, characters 0-29.
+|}]
+
+(***********************************)
+(* Test: Abstracting in signatures *)
+
+(* You can abstract away an alias *)
+module M : sig
+  kind_ k
+
+  type t : k
+
+  val f : unit -> t
+end = struct
+  kind_ k = value
+
+  type t = string
+
+  let f () = "hi mom"
+end
+[%%expect{|
+module M : sig kind_ k type t : k val f : unit -> t end
+|}]
+
+(* But be careful - now we don't know enough to call f! *)
+let _ = M.f ()
+[%%expect{|
+Line 1, characters 8-14:
+1 | let _ = M.f ()
+            ^^^^^^
+Error: This expression has type "M.t" but an expression was expected of type
+         "('a : '_representable_layout_3)"
+       The kind of M.t is M.k
+         because of the definition of t at line 4, characters 2-12.
+       But the kind of M.t must be representable
+         because it's the type of a variable bound by a `let`.
+|}]
+
+(* If you expose the definition, you have to be exact. *)
```

Comment:
It'd be nice to include tests demonstrating why this is necessary via the examples you gave in the other thread.

Reply:
I thought some more about those examples - the signatures I proposed themselves won't typecheck, so I think they are better as an example of why module `with kind_` substitutions must be invariant, and I will add them in the next PR.